sql: fix COPY with some concurrent schema changes

yuzefovich opened the pull request:
Previously, if we had COPY running concurrently with some schema changes
that require usage of Puts (via ForcePut on the index), like ALTER
PRIMARY KEY, we could hit an internal error because the vectorized
encoder didn't support it. This commit adds the support for it, but it
also actually disables using the vectorized encoder for COPY when
ForcePut option is observed on at least one writable index. This is the
case since I observed many different corruption-like failures when
stressing the extended test, which aren't present if we use the
row-by-row insert.

Existing `TestLargeCopy` has been extended so that sometimes it also
performs ALTER PRIMARY KEY concurrently with the COPY as well as
randomized `copy_from_atomic_enabled` value. Note that in some cases
we're hitting "duplicate key" violation while there is a schema change,
and it seems somewhat expected to me, so I didn't look deeper and simply
made the test ignore such errors.

Fixes: #147955.

Release note (bug fix): CockroachDB could previously encounter "vector
encoder doesn't support ForcePut yet" error when performing COPY command
concurrently with some schema changes. The bug has been present since
before 23.2 and is now fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148549)
<!-- Reviewable:end -->

michae2 commented:
[triage] @yuzefovich wondering if this is worth picking up again?

yuzefovich committed:
sql: fix COPY with some concurrent schema changes

Previously, if we had COPY running concurrently with some schema changes
that require usage of Puts (via ForcePut on the index), like ALTER
PRIMARY KEY, we could hit an internal error because the vectorized
encoder didn't support it. This commit adds the support for it, but it
also actually disables using the vectorized encoder for COPY when
ForcePut option is observed on at least one writable index. This is the
case since I observed many different corruption-like failures when
stressing the extended test, which aren't present if we use the
row-by-row insert.

Existing `TestLargeCopy` has been extended so that sometimes it also
performs ALTER PRIMARY KEY concurrently with the COPY as well as
randomized `copy_from_atomic_enabled` value. Note that in some cases
we're hitting "duplicate key" violation while there is a schema change,
and it seems somewhat expected to me, so I didn't look deeper and simply
made the test ignore such errors.

Release note (bug fix): CockroachDB could previously encounter "vector
encoder doesn't support ForcePut yet" error when performing COPY command
concurrently with some schema changes. The bug has been present since
before 23.2 and is now fixed.

yuzefovich commented:
This is RFAL.

An example corruption-like test failure if we do use the vectorized encoder with concurrent ALTER PK:
```
panic: invalid header size: 4
  (1) attached stack trace
    -- stack trace:
    | runtime.gopanic
    |   GOROOT/src/runtime/panic.go:783
    | [...repeated from below...]
  Wraps: (2) attached stack trace
    -- stack trace:
    | github.com/cockroachdb/cockroach/pkg/roachpb.Value.VerifyHeader
    |   pkg/roachpb/data.go:383
    | github.com/cockroachdb/cockroach/pkg/kv.marshalValue
    |   pkg/kv/util.go:47
    | github.com/cockroachdb/cockroach/pkg/kv.(*Batch).put
    |   pkg/kv/batch.go:437
    | github.com/cockroachdb/cockroach/pkg/kv.(*Batch).Put
    |   pkg/kv/batch.go:466
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).constructMergeBatch
    |   pkg/sql/backfill/mvcc_index_merger.go:599
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).merge.func1.1
    |   pkg/sql/backfill/mvcc_index_merger.go:480
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1
    |   pkg/sql/internal.go:1965
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
    |   pkg/sql/internal.go:2052
    | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
    |   pkg/kv/txn.go:1144
    | github.com/cockroachdb/cockroach/pkg/kv.runTxn
    |   pkg/kv/db.go:1101
    | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
    |   pkg/kv/db.go:1064
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func2
    |   pkg/sql/internal.go:2025
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
    |   pkg/sql/internal.go:2039
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
    |   pkg/sql/internal.go:1966
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).merge.func1
    |   pkg/sql/backfill/mvcc_index_merger.go:418
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.retryWithReducedBatchWhenAutoRetryLimitExceeded
    |   pkg/sql/backfill/mvcc_index_merger.go:719
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).merge
    |   pkg/sql/backfill/mvcc_index_merger.go:416
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).Run.func5
    |   pkg/sql/backfill/mvcc_index_merger.go:226
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).Run.Group.GoCtx.func8
    |   pkg/util/ctxgroup/ctxgroup.go:200
    | golang.org/x/sync/errgroup.(*Group).Go.func1
    |   external/org_golang_x_sync/errgroup/errgroup.go:93
    | runtime.goexit
    |   src/runtime/asm_amd64.s:1693
```
I spent little time looking into this but not sure what was going wrong so decided to not use the vectorized engine for COPY when ForcePut is present but still add the support for it.

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Good thing you created that test! Thanks for fixing this!

> @michae2 reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148549#-:-Odlla2J9S_UA1I0lxSg:bt8sos3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/copy_from.go` line 440 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148549#-Odlhc1uEwEZOf9xlDWL:-Odlhc1uEwEZOf9xlDWM:b-b4q99h) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b63b41109f93c9c63f27ee8fd162494bf58cad/pkg/sql/copy_from.go#L440)):*
> ```Go
> 		// COPY with a concurrent ALTER PRIMARY KEY has resulted in different
> 		// corruption scenarios. The non-vectorized COPY doesn't hit those, so
> 		// we choose to fall back.
> ```

Is there already a GH issue open about these corruption scenarios? It might be worth opening one and linking it here with a TODO, since it seems like there are still some problems to figure out.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148549#-:-OdqC8blC7lnw-nKITIO:b-oduykh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

___
*[`pkg/sql/copy_from.go` line 440 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148549#-Odlhc1uEwEZOf9xlDWL:-Odq9arC7lC7lCXDBK8w:b-ukop3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b63b41109f93c9c63f27ee8fd162494bf58cad/pkg/sql/copy_from.go#L440)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is there already a GH issue open about these corruption scenarios? It might be worth opening one and linking it here with a TODO, since it seems like there are still some problems to figure out.
</blockquote></details>

Good point, opened #157198 for this.


<!-- Sent from Reviewable.io -->
